Improve uf

james-d-mitchell opened the pull request.

flsmith commented on include/libsemigroups/uf.hpp:
```diff
@@ -20,70 +20,402 @@
+      static constexpr index_type rank_mask
+          = ((index_type(1) << RANK_BITS) - index_type(1));
+
+      static inline rank_type rank(node_type i) noexcept {
```
assertions on these three functions?

flsmith reviewed (approved):
Looks good! Only some minor points.

james-d-mitchell commented:
Thanks for the review @flsmith, I've modified this as requested, added an attribution to the origin on the implementation, and updated the other comments at the start of the file. If the CI passes, I'll merge.
